README.md
# Fernet Haskell Implementation

[![Build Status](https://travis-ci.org/rvl/fernet-hs.svg?branch=master)](https://travis-ci.org/rvl/fernet-hs) [![Hackage](https://img.shields.io/hackage/v/fernet.svg)](http://hackage.haskell.org/package/fernet)

*Fernet* generates and verifies HMAC-based authentication tokens.

Originally designed for use within OpenStack clusters, it was intended
to be fast and light-weight, with non-persistent tokens. Integrity and
confidentiality of the token contents are implemented with HMAC SHA256
and AES128 CBC.

See the [Fernet Spec][spec] for a little more information.

[spec]: https://github.com/fernet/spec/blob/master/Spec.md

## Usage

To encrypt a token:

    >>> import Network.Fernet
    >>> k <- generateKey
    >>> keyToBase64 k
    "JQAeL3iFN9wIW_hMKiIzA1EiG_EZNivnMPBOOJn2wZc="
    >>> token <- encrypt k "secret text"
    >>> print token
    "gAAAAABY0H9kx7ihkcj6ZF_bQ73Lvc7aG-ZlEtjx24io-DQy5tCjLbq1JvVY27uAe6BuwG8css-4LDIywOJRyY_zetq7aLPPag=="

The resulting token can be distributed to clients. To check and
decrypt the token, use the same key:

    >>> decrypt k 60 token
    Right "secret text"

Do read the [Network.Fernet module][haddock] documentation for further
information.

[haddock]: http://hackage.haskell.org/package/fernet/docs/Network-Fernet.html


## Command-line tool

This package also includes a command-line tool for encrypting and
decrypting tokens.

    Fernet Utility

    Usage: fernet (((-k|--key STRING) | --key-file FILENAME) ([-e|--encrypt] |
                  [-d|--decrypt]) [--ttl SECONDS] | (-g|--gen-key))
      Encrypts/decrypts Fernet tokens. One token written to stdout for each line
      read from stdin. Use --gen-key to make a key.

    Available options:
      -h,--help                Show this help text
      -k,--key STRING          Base64-urlsafe-encoded 32 byte encryption key
      --key-file FILENAME      File containing the encryption key
      -e,--encrypt             Encryption mode (default: autodetect)
      -d,--decrypt             Decryption mode (default: autodetect)
      --ttl SECONDS            Token lifetime in seconds (default: 1 minute)
      -g,--gen-key             Generate a key from the password on standard input

## Development

### Building with Stack

```
stack build
```

### Building with Nix

```
nix-shell -p cabal2nix --command "cabal2nix --shell . > default.nix"
nix-shell --command "cabal configure"
cabal build
```

## Better & Cooler Stuff

You might also be interested in [hsoz](https://github.com/rvl/hsoz).


ChangeLog.md
# Revision history for fernet

## 0.1.0.0  -- 2017-03-22

* First version. Released on an unsuspecting world.


test/Main.hs
module Main where

import           Test.Tasty (defaultMain, testGroup)

import qualified Network.Fernet.Tests

main :: IO ()
main = do
  tests <- Network.Fernet.Tests.makeTests
  defaultMain $ testGroup "Tests" [ tests ]


test/Network/Fernet/Tests.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}
{-# LANGUAGE DeriveGeneric #-}

module Network.Fernet.Tests (makeTests) where

import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Char8 as S8
import Data.Maybe (fromJust, fromMaybe)
import Data.Either (isLeft)
import Data.Aeson
import Control.Applicative
import Data.Time.Clock (NominalDiffTime)
import Data.Time.LocalTime (zonedTimeToUTC)
import Data.Time.Format (parseTimeM, defaultTimeLocale)
import Data.Time.Clock.POSIX (POSIXTime, utcTimeToPOSIXSeconds)
import GHC.Generics
import System.IO

import Test.QuickCheck
import Test.QuickCheck.Monadic
import qualified Test.QuickCheck         as QC
import qualified Test.QuickCheck.Monadic as QC
import           Test.Tasty              (TestTree, testGroup)
import           Test.Tasty.HUnit        (testCase)
import           Test.HUnit              (Assertion, (@?=), assertFailure)

import Network.Fernet

data Spec = Spec
            { desc :: String
            , token :: ByteString
            , now :: POSIXTime
            , ttl :: NominalDiffTime
            , secret :: ByteString
            , src :: ByteString
            , iv :: ByteString
            } deriving (Generic, Show)

instance FromJSON Spec where
    parseJSON = withObject "Spec" $ \v -> Spec
        <$> v .:? "desc" .!= ""
        <*> liftA S8.pack (v .: "token")
        <*> liftA (fromMaybe 0 . parseTime) (v .: "now")
        <*> liftA (fromIntegral :: Int -> NominalDiffTime) (v .:? "ttl_sec" .!= 0)
        <*> liftA S8.pack (v .: "secret")
        <*> liftA S8.pack (v .:? "src" .!= "")
        <*> liftA BS.pack (v .:? "iv" .!= [])

parseTime :: Monad m => String -> m POSIXTime
parseTime = fmap (utcTimeToPOSIXSeconds . zonedTimeToUTC) . parseTimeM False defaultTimeLocale fmt
  where fmt = "%Y-%m-%dT%H:%M:%S%z"

makeTests :: IO TestTree
makeTests = do
  generate <- makeGroup "generate" generateTest
  verify <- makeGroup "verify" verifyTest
  invalid <- makeGroup "invalid" invalidTest
  return $ testGroup "Acceptance Tests" [ generate, verify, invalid ]

makeGroup :: String -> (Spec -> TestTree) -> IO TestTree
makeGroup name makeSpec = do
  let f = "spec/" ++ name ++ ".json"
  withFile f ReadMode $ \h -> do
    c <- BL.hGetContents h
    case eitherDecode' c of
      Right specs -> return $ testGroup name (map makeSpec specs)
      Left e -> return $ testGroup name [testCase ("Loading " ++ f) (assertFailure e)]

getKey :: ByteString -> IO Key
getKey secret = case keyFromBase64 secret of
                  Right key -> return key
                  Left e -> fail $ "Couldn't decode secret: " ++ e

generateTest :: Spec -> TestTree
generateTest Spec{..} = testCase "Token generation" $ do
  key <- getKey secret
  encrypt' key now iv src @?= token

verifyTest :: Spec -> TestTree
verifyTest Spec{..} = testCase "Successful token verification" $ do
  key <- getKey secret
  decrypt' key ttl now token @?= Right src

invalidTest :: Spec -> TestTree
invalidTest Spec{..} = testCase desc $ do
  key <- getKey secret
  let r = decrypt' key ttl now token
  isLeft r @?= True
  


src/Network/Fernet.hs
-- | /Fernet/ generates and verifies HMAC-based authentication tokens.
--
-- Originally designed for use within OpenStack clusters, it was
-- intended to be fast and light-weight, with non-persistent
-- tokens. Integrity and confidentiality of the token contents are
-- implemented with HMAC SHA256 and AES128 CBC.
--
-- See the <https://github.com/fernet/spec/blob/master/Spec.md Fernet Spec>
-- for a little more information.
--
-- == Usage
-- To encrypt a token:
--
-- >>> import Network.Fernet
-- >>> k <- generateKey
-- >>> keyToBase64 k
-- "JQAeL3iFN9wIW_hMKiIzA1EiG_EZNivnMPBOOJn2wZc="
-- >>> token <- encrypt k "secret text"
-- >>> print token
-- "gAAAAABY0H9kx7ihkcj6ZF_bQ73Lvc7aG-ZlEtjx24io-DQy5tCjLbq1JvVY27uAe6BuwG8css-4LDIywOJRyY_zetq7aLPPag=="
--
-- The resulting token can be distributed to clients. To check and
-- decrypt the token, use the same key:
--
-- >>> decrypt k 60 token
-- Right "secret text"
--
-- When decrypting, a TTL value is supplied to determine whether the
-- token has expired. The timestamp is stored in plain text and can
-- also be checked with 'hasExpired'.
--
-- == Related Modules
--
-- * "Network.Iron"
-- * "Jose.Jwt"

module Network.Fernet
  ( -- * Tokens
    encrypt
  , decrypt
  , encrypt'
  , decrypt'
  , DecryptError(..)
  , isExpired
  , hasExpired
  -- * Keys
  , Key
  , key
  , generateKey
  , generateKeyFromPassword
  , keyFromBase64
  , keyToBase64
  -- * Other
  , version
  ) where

import           Data.ByteString        (ByteString)
import qualified Data.ByteString        as BS
import           Data.ByteArray         (ScrubbedBytes)
import           Data.Byteable          (constEqBytes)
import           Data.Word              (Word8)
import           Data.Time.Clock        (NominalDiffTime)
import           Data.Time.Clock.POSIX  (POSIXTime, getPOSIXTime)
import           Data.Bifunctor         (first)

import Network.Fernet.Crypto
import Network.Fernet.Key
import Network.Fernet.Token

-- | @0x80@ is the latest token format version, and the only one
-- supported by this library.
version :: Word8
version = 0x80

----------------------------------------------------------------------------
-- Encryption

-- | Encrypts, encodes, and signs the given token contents with the
-- given key.
--
-- Its timestamp is set to the current time and stored /unencrypted/
-- in the token.
encrypt :: Key -- ^ The encryption and signing keys.
        -> ByteString -- ^ Token contents.
        -> IO ByteString -- ^ An encoded /Fernet/ token.
encrypt k text = do
  ts <- getPOSIXTime
  iv <- genIV
  return $ encrypt' k ts iv text

-- | Encrypts, encodes, and signs the given token contents with the
-- given key.
--
-- The provided timestamp is stored /unencrypted/ in the token.
--
-- The given IV (initialization vector) string should be a random
-- sequence of exactly 128 bits.
encrypt' :: Key          -- ^ The encryption and signing keys.
         -> POSIXTime    -- ^ Timestamp
         -> ByteString   -- ^ Initialization Vector.
         -> ByteString   -- ^ Token contents.
         -> ByteString   -- ^ An encoded /Fernet/ token.
encrypt' Key{..} ts iv text =
  case serialize <$> makeToken encryptionKey ts iv text of
    Just token -> encode token (sign signingKey token)
    Nothing -> "" -- this shouldn't happen, unless iv is wrong

makeToken :: ScrubbedBytes     -- ^ Keys
          -> POSIXTime         -- ^ Timestamp
          -> ByteString        -- ^ Initialization Vector
          -> ByteString        -- ^ Plain text
          -> Maybe TokenFields
makeToken k ts iv text = TokenFields version ts iv <$> ct
  where ct = aesEncrypt k iv text

----------------------------------------------------------------------------
-- Decryption

-- | Some of the reasons why decryption can fail.
data DecryptError = TokenMalformed     -- ^ The token could not be decoded into fields.
                  | TokenInvalid       -- ^ Signature verification failed.
                  | TokenExpired       -- ^ Token age exceeded given TTL value.
                  | UnacceptableClockSkew -- ^ Token timestamp is too far in the future.
                  | KeySizeInvalid     -- ^ The key was not suitable for decryption.
                  | InvalidBlockSize   -- ^ The ciphertext length was not a multiple of the block size.
                  | UnsupportedVersion -- ^ The version was not 0x80.
                  deriving (Show, Eq)

-- | Decodes, checks, and decrypts, the given /Fernet/ token.
--
-- If the token's age (determined by its timestamp) exceeds the given
-- TTL, then this function will fail.
decrypt :: Key             -- ^ The encryption and signing keys.
        -> NominalDiffTime -- ^ Token TTL.
        -> ByteString      -- ^ The encoded token.
        -> IO (Either DecryptError ByteString) -- ^ Token contents, or an error.
decrypt k ttl t = do
  now <- getPOSIXTime
  return $ decrypt' k ttl now t

-- | Decodes, checks, and decrypts, the given /Fernet/ token.
--
-- If the token's age (determined by its timestamp) exceeds the given
-- TTL, then this function will fail.
decrypt' :: Key             -- ^ The encryption and signing keys.
         -> NominalDiffTime -- ^ Token TTL.
         -> POSIXTime       -- ^ The current time, used to determine token age.
         -> ByteString      -- ^ The encoded token.
         -> Either DecryptError ByteString -- ^ Token contents, or an error.
decrypt' Key{..} ttl now t = do
  (fields, tb, sig) <- first (const TokenMalformed) (decode t)
  checkVersion fields
  checkTimestamp now fields
  checkExpiry ttl now fields
  checkSignature signingKey tb sig
  checkInputSize fields
  case aesDecrypt encryptionKey (tokenIV fields) (tokenCiphertext fields) of
    Just text -> Right text
    Nothing -> Left KeySizeInvalid

checkVersion :: TokenFields -> Either DecryptError ()
checkVersion tf | tokenVersion tf == version = Right ()
                | otherwise = Left UnsupportedVersion

-- | Maximum clock skew in the future direction.
maxClockSkew :: NominalDiffTime
maxClockSkew = 60

checkTimestamp :: POSIXTime -> TokenFields -> Either DecryptError ()
checkTimestamp now TokenFields{..} | tokenTimestamp - now <= maxClockSkew = Right ()
                                   | otherwise = Left UnacceptableClockSkew

checkExpiry :: NominalDiffTime -> POSIXTime -> TokenFields -> Either DecryptError ()
checkExpiry ttl now tf | hasExpired' ttl now tf = Right ()
                       | otherwise = Left TokenExpired

checkSignature :: ScrubbedBytes -> ByteString -> ByteString -> Either DecryptError ()
checkSignature k tf sig | constEqBytes sig (sign k tf) = Right ()
                        | otherwise                    = Left TokenInvalid

checkInputSize :: TokenFields -> Either DecryptError ()
checkInputSize tf | isBlocked (tokenCiphertext tf) = Right ()
                  | otherwise                      = Left InvalidBlockSize
  where isBlocked t = BS.length t `mod` cipherBlockSize == 0


src/Network/Fernet/Key.hs
module Network.Fernet.Key
  ( Key(..)
  , key
  , generateKey
  , generateKeyFromPassword
  , keyFromBase64
  , keyToBase64
  ) where

import           Data.Monoid            ((<>))
import           Data.ByteString        (ByteString)
import qualified Data.ByteString        as BS
import           Data.Byteable          (Byteable(..))
import           Data.ByteArray         (ScrubbedBytes, ByteArrayAccess(..))
import qualified Data.ByteArray         as BA
import qualified Crypto.KDF.PBKDF2      as PBKDF2
import           Crypto.Hash.Algorithms (SHA256(..))
import           Crypto.Random          (getRandomBytes)

import Network.Fernet.Base64

-- | Contains the signing key and encryption key. Create one with
-- 'key', 'keyFromBase64', or 'generateKeyFromPassword'.
data Key = Key
           { signingKey    :: ScrubbedBytes
           , encryptionKey :: ScrubbedBytes
           } deriving (Show, Eq)

-- | Constructs a pair of signing and encryption keys. Each key must
-- be exactly 16 bytes long or this will fail.
key :: ByteArrayAccess a
    => a   -- ^ Signing Key
    -> a   -- ^ Encryption Key
    -> Maybe Key
key s e = Key <$> toKey checkHashKeyLength s <*> toKey checkCipherKeyLength e

toKey :: ByteArrayAccess a => (Int -> Bool) -> a -> Maybe ScrubbedBytes
toKey checkLength k | checkLength (BA.length k) = Just (BA.convert k)
                    | otherwise                      = Nothing

cipherKeyLength :: Int
cipherKeyLength = 16

-- | Check that key length is appropriate for AES128.
checkCipherKeyLength :: Int -> Bool
checkCipherKeyLength = (== cipherKeyLength)

checkHashKeyLength :: Int -> Bool
checkHashKeyLength = (>= 16)

-- | Generates new keys from the PRNG.
generateKey :: IO Key
generateKey = splitKeys <$> getRandomBytes (cipherKeyLength * 2)

-- | Input must be exactly length 32 chars
splitKeys :: ByteString -> Key
splitKeys = make . BS.splitAt cipherKeyLength
  where make (s, e) = Key (BA.convert s) (BA.convert e)

genSalt :: IO ByteString
genSalt = getRandomBytes 16

-- | Encodes the given key as urlsafe base64.
keyToBase64 :: Key -> ByteString -- ^ URL-safe base64.
keyToBase64 (Key s e) = b64url $ s <> e

-- | Decodes urlsafe base64-encoded bytes into a key. This will fail
-- if the input is not exactly 256 bits long (43 characters in
-- base64).
keyFromBase64 :: ByteString -- ^ URL-safe base64.
              -> Either String Key
keyFromBase64 = (>>= make) . b64urldec
  where make s = case key sk ek of
                   Just k -> Right k
                   Nothing -> Left "Invalid key length"
          where (sk, ek) = BS.splitAt ((BS.length s) - 16) s

-- | Stretches the given password into a 'Key' using PBKDF2.
generateKeyFromPassword :: Byteable p
                        => Int -- ^ Number of key derivation function iterations.
                        -> p   -- ^ The password.
                        -> IO (Key, ByteString) -- ^ The key and random salt used.
generateKeyFromPassword iterations p = do
  salt <- genSalt
  let keys = PBKDF2.generate prf params (toBytes p) salt
  return (splitKeys keys, salt)
  where
    prf = PBKDF2.prfHMAC SHA256
    params = PBKDF2.Parameters iterations 32


src/Network/Fernet/Token.hs
module Network.Fernet.Token
  ( encode
  , decode
  , serialize
  , deserialize
  , isExpired
  , hasExpired
  , hasExpired'
  , TokenFields(..)
  , Signature
  ) where

import           Data.ByteString        (ByteString)
import qualified Data.ByteString        as BS
import qualified Data.ByteString.Lazy   as BL
import Data.Word (Word8)
import Data.Time.Clock.POSIX (POSIXTime, getPOSIXTime)
import Data.Time.Clock (NominalDiffTime)
import Data.Binary.Get
import Data.Binary.Put

import Network.Fernet.Base64

data TokenFields = TokenFields
  { tokenVersion    :: Word8      -- ^ Version, 8 bits
  , tokenTimestamp  :: POSIXTime  -- ^ Timestamp, 64 bits
  , tokenIV         :: ByteString -- ^ IV, 128 bits
  , tokenCiphertext :: ByteString -- ^ Ciphertext, variable length, multiple of 128 bits
  } deriving (Show, Eq)

type Signature = ByteString

-- | Size of a SHA256 hash.
hmacLength :: Int
hmacLength = 32

encode :: ByteString -> Signature -> ByteString
encode t s = b64url $ BS.concat [t, s]

decode :: ByteString -> Either String (TokenFields, ByteString, Signature)
decode = (>>= decode') . b64urldec
  where
    decode' bs = do
      (t, sig) <- splitToken bs
      tf <- deserialize t
      return (tf, t, sig)
    splitToken bs | BS.length sig < hmacLength = Left "Missing HMAC"
                  | otherwise = Right (t, sig)
      where (t, sig) = BS.splitAt (BS.length bs - hmacLength) bs

serialize :: TokenFields -> ByteString
serialize TokenFields{..} = BL.toStrict . runPut $ do
  putWord8 tokenVersion
  putWord64be (floor tokenTimestamp)
  putByteString tokenIV
  putByteString tokenCiphertext

deserialize :: ByteString -> Either String TokenFields
deserialize t = case runGetOrFail get (BL.fromStrict t) of
                  Left (_, _, e) -> Left e
                  Right (_, _, tf) -> Right tf
  where get = do
          v <- getWord8
          ts <- getWord64be
          iv <- getByteString 16
          ct <- BL.toStrict <$> getRemainingLazyByteString
          return $! TokenFields v (fromIntegral ts) iv ct

-- | Returns @Right True@ if the token has expired,
-- @Left _@ if the token could not be parsed.
hasExpired :: NominalDiffTime -- ^ TTL value.
           -> ByteString      -- ^ Encoded token.
           -> IO (Either String Bool)
hasExpired ttl token = isExpired ttl token <$> getPOSIXTime

-- | Returns @Right True@ if the token is expired at the given time,
-- @Left _@ if the token could not be parsed.
isExpired :: NominalDiffTime -- ^ TTL value.
          -> ByteString      -- ^ Encoded token.
          -> POSIXTime       -- ^ The time to consider.
          -> Either String Bool
isExpired ttl token now = do
  (tf, _, _) <- decode token
  return $ hasExpired' ttl now tf

hasExpired' :: NominalDiffTime -> POSIXTime -> TokenFields -> Bool
hasExpired' ttl now TokenFields{..} = now - tokenTimestamp < ttl


src/Network/Fernet/Base64.hs
-- | Base64 utilities

module Network.Fernet.Base64 (
  -- * Base64
    b64
  , b64dec
  , b64url
  , b64urldec
  ) where

import Data.Monoid ((<>))
import qualified Data.ByteString.Char8   as S8
import           Data.ByteArray          (ByteArrayAccess)
import qualified Data.ByteArray.Encoding as B (Base (..), convertToBase, convertFromBase)
import           Data.ByteString         (ByteString)

-- | Shorthand for encode in Base64.
b64 :: ByteArrayAccess a => a -> ByteString
b64 = B.convertToBase B.Base64

b64url :: ByteArrayAccess a => a -> ByteString
b64url = urlSafeBase64 . b64

b64dec :: ByteArrayAccess a => a -> Either String ByteString
b64dec = B.convertFromBase B.Base64

b64urldec :: ByteString -> Either String ByteString
b64urldec = b64dec . unUrlSafeBase64

-- | Fixes up a Base64 encoded string so that it's more convenient to
-- include in URLs.
-- The characters @+@ and @/@ are replaced with @-@ and @_@.
urlSafeBase64 :: ByteString -> ByteString
urlSafeBase64 = S8.map (tr '+' '-' . tr '/' '_')

-- | The inverse of 'urlSafeBase64'.
unUrlSafeBase64 :: ByteString -> ByteString
unUrlSafeBase64 = S8.map (tr '-' '+' . tr '_' '/')

tr :: Char -> Char -> Char -> Char
tr a b c = if c == a then b else c


src/Network/Fernet/Crypto.hs
module Network.Fernet.Crypto
  ( sign
  , aesEncrypt
  , aesDecrypt
  , genIV
  , cipherBlockSize
  ) where

import           Data.ByteString        (ByteString)
import           Data.ByteArray         (ByteArray, ByteArrayAccess, convert)

import           Crypto.Data.Padding    (Format(PKCS7), pad, unpad)
import           Crypto.Hash.Algorithms (SHA256 (..))
import           Crypto.Cipher.AES      (AES128)
import           Crypto.MAC.HMAC        (HMAC(..), hmac, hmacGetDigest)
import           Crypto.Cipher.Types
import           Crypto.Error
import           Crypto.Random          (getRandomBytes)

import Network.Fernet.Token (Signature)

sign :: ByteArrayAccess a => a -> ByteString -> Signature
sign key t = convert $ hmacGetDigest (hmac key t :: HMAC SHA256)

aesEncrypt :: ByteArray a
           => a          -- ^ The encryption key
           -> ByteString -- ^ Initialization Vector
           -> ByteString -- ^ Plain text
           -> Maybe ByteString
aesEncrypt key iv text = cbcEncrypt <$> ctx <*> iv' <*> text'
  where
    ctx = maybeCryptoError (cipherInit key) :: Maybe AES128
    iv' = makeIV iv
    p = fmap (PKCS7 . blockSize) ctx
    text' = pad <$> p <*> pure text

aesDecrypt :: ByteArray a
           => a           -- ^ The encryption key
           -> ByteString -- ^ Initialization Vector
           -> ByteString -- ^ Cipher text
           -> Maybe ByteString
aesDecrypt key iv ct = do
  (ctx, iv', p) <- aesSetup key iv
  let text' = cbcDecrypt ctx iv' ct
  unpad p text'

-- | Block size for AES128
cipherBlockSize :: Int
cipherBlockSize = 16

genIV :: IO ByteString
genIV = getRandomBytes cipherBlockSize

aesSetup :: ByteArray a => a -> ByteString -> Maybe (AES128, IV AES128, Format)
aesSetup key iv = (,,) <$> ctx <*> iv' <*> p
  where
    ctx = maybeCryptoError (cipherInit key)
    iv' = makeIV iv
    p = PKCS7 . blockSize <$> ctx


cli/FernetMain.hs
module Main where

import Options.Applicative
import Data.Monoid ((<>))
import Control.Monad (join, unless, when)
import Data.Time.Clock (NominalDiffTime)
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Char8 as S8
import qualified Data.ByteString.Lazy.Char8 as L8
import qualified Data.ByteArray.Encoding as B (Base(..), convertFromBase)
import Text.Read (readEither)
import System.IO
import Data.Bifunctor (first)
import System.Posix.IO (handleToFd, stdInput)
import System.Posix.Terminal (queryTerminal)

import Network.Fernet

main :: IO ()
main = join . execParser $
  info (helper <*> parser)
  (  fullDesc
  <> header "Fernet Utility"
  <> progDesc (
      "Encrypts/decrypts Fernet tokens. " ++
      "One token written to stdout for each line read from stdin. " ++
      "Use --gen-key to make a key."
  ))
  where
    parser :: Parser (IO ())
    parser =
      (fernet
        <$> (KeyText <$> ( strOption
              ( long "key"
                <> short 'k'
                <> metavar "STRING"
                <> help "Base64-urlsafe-encoded 32 byte encryption key"
              )) <|>
              KeyFile <$> ( strOption
                ( long "key-file"
                  <> metavar "FILENAME"
                  <> help "File containing the encryption key"
                )
              ))
        <*> ( optional
              ( flag' Encrypt
                ( long "encrypt"
                  <> short 'e'
                  <> help "Encryption mode (default: autodetect)" )
              <|>
              flag' Decrypt
                ( long "decrypt"
                  <> short 'd'
                  <> help "Decryption mode (default: autodetect)" )
              )
            )
        <*> option ttl
            ( long "ttl"
            <> metavar "SECONDS"
            <> help "Token lifetime in seconds (default: 1 minute)"
            <> value 60
            )) <|>
      (genKey <$> ( flag' True
                    (long "gen-key"
                    <> short 'g'
                    <> help "Generate a key from the password on standard input"
                    )))

genKey :: Bool -> IO ()
genKey _ = do
  password <- askPassword
  (k, _) <- generateKeyFromPassword iterations password
  S8.hPutStrLn stdout (keyToBase64 k)

askPassword ::  IO ByteString
askPassword = do
  isatty <- queryTerminal stdInput
  when isatty $ do
    hSetEcho stdin False
    S8.hPutStr stderr "Enter password: "
    hFlush stdout
  password <- S8.hGetLine stdin
  when isatty $ do
    hSetEcho stdin True
    S8.hPut stderr "\n"
  return password

ttl :: ReadM NominalDiffTime
ttl = eitherReader (fmap fromInteger . readEither)

data Action = Encrypt | Decrypt
data Keys = KeyText String | KeyFile FilePath
data Password = PasswordText String | PasswordFile FilePath

fernet :: Keys -> Maybe Action -> NominalDiffTime -> IO ()
fernet ks ax ttl = do
  k <- readKeys ks
  L8.hGetContents stdin >>= mapM_ (processLine k ax ttl) . L8.lines

processLine :: Key -> Maybe Action -> NominalDiffTime -> L8.ByteString -> IO ()
processLine k ax ttl s = doLine k ax ttl s >>= uncurry L8.hPutStrLn . output

output :: Either String ByteString -> (Handle, L8.ByteString)
output (Left e)  = (stderr, L8.pack e)
output (Right s) = (stdout, L8.fromStrict s)

doLine :: Key -> Maybe Action -> NominalDiffTime -> L8.ByteString -> IO (Either String ByteString)
doLine k (Just Encrypt) _   s = lineEncrypt k s
doLine k (Just Decrypt) ttl s = lineDecrypt k ttl s
doLine k Nothing        ttl s = doLine k (Just $ sniff s) ttl s

sniff :: BL.ByteString -> Action
sniff s | ver >= "gA" && ver <= "gP" = Decrypt
        | otherwise                  = Encrypt
  where ver = BL.take 2 s

lineEncrypt :: Key -> L8.ByteString -> IO (Either String ByteString)
lineEncrypt k s = Right <$> encrypt k (L8.toStrict s)

lineDecrypt :: Key -> NominalDiffTime -> L8.ByteString -> IO (Either String ByteString)
lineDecrypt k ttl s = first show <$> decrypt k ttl (L8.toStrict s)

readKeys :: Keys -> IO Key
readKeys (KeyText k) = keyFromString k
readKeys (KeyFile f) = readFirstLine f >>= \k -> readKeys (KeyText k)

readPassword :: Password -> IO Key
readPassword (PasswordText p) = fst <$> generateKeyFromPassword iterations (S8.pack p)
readPassword (PasswordFile f) = readFirstLine f >>= \p -> readPassword (PasswordText p)

iterations = 100000 :: Int

readFirstLine :: FilePath -> IO String
readFirstLine f = withFile f ReadMode hGetLine

keyFromString :: String -> IO Key
keyFromString s = case keyFromBase64 (S8.pack s) of
                    Right k -> return k
                    Left e -> fail e

-- | Converts 'Maybe' to 'Either'.
justRight :: e -> Maybe a -> Either e a
justRight _ (Just a) = Right a
justRight e Nothing = Left e
